halo2/src/prover.rs

```diff
@@ -26,178 +26,155 @@ use crate::circuit_builder::analyzed_to_circuit;
 
 use itertools::Itertools;
 use rand::rngs::OsRng;
-use std::io;
-use std::io::Cursor;
-use std::time::Instant;
+use std::{
+    io::{self, Cursor},
+    time::Instant,
+};
+
+pub use halo2_proofs::halo2curves::bn256::Bn256;
+pub use halo2_proofs::poly::kzg::commitment::ParamsKZG;
 
 /// Create a halo2 proof for a given PIL, fixed column values and witness column values
 /// We use KZG ([GWC variant](https://eprint.iacr.org/2019/953)) and Keccak256
+pub struct Halo2Prover {
+    params: ParamsKZG<Bn256>,
+}
 
-pub fn prove_ast_read_params<T: FieldElement, R: io::Read>(
-    pil: &Analyzed<T>,
-    fixed: Vec<(&str, Vec<T>)>,
-    witness: Vec<(&str, Vec<T>)>,
-    mut params: R,
-) -> Vec<u8> {
-    if polyexen::expr::get_field_p::<Fr>() != T::modulus().to_arbitrary_integer() {
-        panic!("powdr modulus doesn't match halo2 modulus. Make sure you are using Bn254");
+impl Halo2Prover {
+    pub fn new(size: DegreeType) -> Self {
+        let degree = DegreeType::BITS - size.leading_zeros() + 1;
+        Self {
+            params: ParamsKZG::<Bn256>::new(degree),
+        }
     }
 
-    let params = ParamsKZG::<Bn256>::read(&mut params).unwrap();
-    prove_ast(pil, fixed, witness, params)
-}
+    pub fn new_from_setup(input: &mut impl io::Read) -> Result<Self, io::Error> {
+        let params = ParamsKZG::<Bn256>::read(input)?;
 
-pub fn prove_ast<T: FieldElement>(
-    pil: &Analyzed<T>,
-    fixed: Vec<(&str, Vec<T>)>,
-    witness: Vec<(&str, Vec<T>)>,
-    params: ParamsKZG<Bn256>,
-) -> Vec<u8> {
-    if polyexen::expr::get_field_p::<Fr>() != T::modulus().to_arbitrary_integer() {
-        panic!("powdr modulus doesn't match halo2 modulus. Make sure you are using Bn254");
+        Ok(Self { params })
     }
 
-    // TODO this is hacky
-    let degree = usize::BITS - fixed[0].1.len().leading_zeros() + 1;
-    let params = {
-        let mut params = params;
-        params.downsize(degree);
-        params
-    };
+    pub fn write(&self, output: &mut impl io::Write) -> Result<(), io::Error> {
```

Comment:
Is this supposed to be `write_setup` or just `write`? Either fine with me

Reply:
yes, it was supposed to be `write_setup`